File: solutions/python/Swap Nodes in Pairs.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None;
        if not head.next:
            return head;
        stack=[]
        last_node=None
        temp=None
        while head:
            temp=head.next;
            head.next=None;
            stack.append(head)
            head=temp;
        if len(stack)%2!=0:
            last_node=stack.pop()
        while len(stack)>0:
            node1=stack.pop()
            node2=stack.pop()
            if temp==None:
                node1.next=node2;
                temp=node1;
            else:
                node2.next=temp;
                node1.next=node2;
                temp=node1;
        if last_node:
            current=temp;
            while current.next:
                current=current.next;
            current.next=last_node;
        return temp;

File: solutions/cpp/Swap Nodes in Pairs.cpp
#include <stack>
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *swapPairs(ListNode *head)
    {
        if (!head)
            return NULL;
        if (!head->next)
            return head;
        std::stack<ListNode *> s;
        ListNode *last_node = NULL;
        ListNode *temp = NULL;
        while (head)
        {
            ListNode *temp = head->next;
            head->next = NULL;
            s.push(head);
            head = temp;
        }
        if (s.size() % 2 != 0)
        {
            last_node = s.top();
            s.pop();
        }
        while (s.size() > 0)
        {
            ListNode *node1 = s.top();
            s.pop();
            ListNode *node2 = s.top();
            s.pop();
            if (temp == NULL)
            {
                node1->next = node2;
                temp = node1;
            }
            else
            {
                node2->next = temp;
                node1->next = node2;
                temp = node1;
            }
        }
        if (last_node)
        {
            ListNode *current = temp;
            while (current->next)
                current = current->next;
            current->next = last_node;
        }
        return temp;
    }
};